File: Harvest/ViewModels/SafetyDashboardViewModel.swift
import Foundation
import Observation

@Observable
final class SafetyDashboardViewModel {
    var analyses: [SafetyAnalysis] = []
    var selectedAnalysis: SafetyAnalysis?
    var redFlags: [RedFlagReport] = []
    var profiles: [String: UserProfile] = [:]
    var isLoading = false
    var error: String?

    private let safetyService = SafetyAnalysisService()
    private let profileService = ProfileService()

    func loadDashboard(userId: String) async {
        isLoading = true
        defer { isLoading = false }

        do {
            analyses = try await safetyService.getSafetyDashboard(userId: userId)

            // Load profiles for each analysis
            for analysis in analyses {
                if profiles[analysis.otherUserId] == nil {
                    if let profile = try? await profileService.getProfile(userId: analysis.otherUserId) {
                        profiles[analysis.otherUserId] = profile
                    }
                }
            }
        } catch {
            self.error = error.localizedDescription
        }
    }

    func loadRedFlags(analysisId: String) async {
        do {
            redFlags = try await safetyService.getRedFlags(analysisId: analysisId)
        } catch {
            self.error = error.localizedDescription
        }
    }

    // MARK: - Retroactive Analysis

    func runBulkRetroactiveAnalysis(userId: String) async {
        isLoading = true
        error = nil
        defer { isLoading = false }

        do {
            let count = try await safetyService.analyzeAllUserConversations(userId: userId)
            print("Bulk retroactive analysis complete: \(count) conversations analyzed")

            // Reload dashboard with updated data
            await loadDashboard(userId: userId)

        } catch {
            self.error = "Failed to analyze conversations: \(error.localizedDescription)"
            print("Error during bulk retroactive analysis: \(error)")
        }
    }
}

File: android/app/src/main/java/com/harvestglass/harvest/ui/safety/SafetyDashboardViewModel.kt
package com.harvestglass.harvest.ui.safety

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.harvestglass.harvest.data.model.RedFlagReport
import com.harvestglass.harvest.data.model.SafetyAnalysis
import com.harvestglass.harvest.data.model.UserProfile
import com.harvestglass.harvest.data.service.ProfileService
import com.harvestglass.harvest.data.service.SafetyAnalysisService
import com.harvestglass.harvest.util.userMessage
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

data class SafetyDashboardUiState(
    val analyses: List<SafetyAnalysis> = emptyList(),
    /** userId → profile, for the names beside each score. */
    val profiles: Map<String, UserProfile> = emptyMap(),
    val selectedAnalysis: SafetyAnalysis? = null,
    val redFlags: List<RedFlagReport> = emptyList(),
    val isLoading: Boolean = false,
    val isAnalyzing: Boolean = false,
    val error: String? = null
)

/** Mirrors Harvest/ViewModels/SafetyDashboardViewModel.swift. */
@HiltViewModel
class SafetyDashboardViewModel @Inject constructor(
    private val safety: SafetyAnalysisService,
    private val profileService: ProfileService
) : ViewModel() {

    private val _state = MutableStateFlow(SafetyDashboardUiState())
    val state: StateFlow<SafetyDashboardUiState> = _state.asStateFlow()

    fun load(userId: String) = viewModelScope.launch {
        _state.update { it.copy(isLoading = true) }
        try {
            val analyses = safety.getSafetyDashboard(userId)
            _state.update { it.copy(analyses = analyses, error = null) }
            loadProfiles(analyses.map { it.otherUserId }.filter { it.isNotEmpty() }.toSet())
        } catch (e: Exception) {
            _state.update { it.copy(error = e.userMessage()) }
        } finally {
            _state.update { it.copy(isLoading = false) }
        }
    }

    fun select(analysis: SafetyAnalysis) = viewModelScope.launch {
        _state.update { it.copy(selectedAnalysis = analysis, redFlags = emptyList()) }
        val flags = runCatching { safety.getRedFlags(analysis.id) }.getOrNull().orEmpty()
        _state.update { it.copy(redFlags = flags) }
    }

    fun clearSelection() = _state.update { it.copy(selectedAnalysis = null, redFlags = emptyList()) }

    fun clearError() = _state.update { it.copy(error = null) }

    /**
     * Re-scores every conversation. Threads that predate the scoring have no
     * flags at all until this runs.
     */
    fun runBulkAnalysis(userId: String) = viewModelScope.launch {
        _state.update { it.copy(isAnalyzing = true) }
        try {
            safety.analyzeAllUserConversations(userId)
            load(userId).join()
        } catch (e: Exception) {
            _state.update { it.copy(error = e.userMessage()) }
        } finally {
            _state.update { it.copy(isAnalyzing = false) }
        }
    }

    private suspend fun loadProfiles(userIds: Set<String>) {
        if (userIds.isEmpty()) return

        val loaded = coroutineScope {
            userIds
                .map { id -> async { runCatching { profileService.getProfile(id) }.getOrNull() } }
                .mapNotNull { it.await() }
                .associateBy { it.id }
        }

        _state.update { it.copy(profiles = it.profiles + loaded) }
    }
}
